core/src/test/scala/unit/kafka/coordinator/AbstractCoordinatorConcurrencyTest.scala

```diff
@@ -201,8 +201,8 @@ object AbstractCoordinatorConcurrencyTest {
         }
       }
       val producerRequestKeys = entriesPerPartition.keys.map(TopicPartitionOperationKey(_)).toSeq
-      watchKeys ++= producerRequestKeys
       producePurgatory.tryCompleteElseWatch(delayedProduce, producerRequestKeys)
+      watchKeys ++= producerRequestKeys
```

Comment:
@junrao This change avoids deadlock in ```TransactionCoordinatorConcurrencyTest```.

If we update ```watchKeys``` before ```tryCompleteElseWatch```, the other threads can take the same key to complete delayed request.  Hence the deadlock happens due to following conditions.

**thread_1**  holds ```stateLock``` of TransactionStateManager to call ```appendRecords``` and it requires lock of delayed request to call ```tryCompleteElseWatch```.

**thread_2** holds lock of delayed request to call ```onComplete``` (updateCacheCallback) and ```updateCacheCallback``` requires ```stateLock``` of TransactionStateManager.

Reply:
According to above case, there is a potential deadlock.

```
    var watchCreated = false
    for(key <- watchKeys) {
      // If the operation is already completed, stop adding it to the rest of the watcher list.
      if (operation.isCompleted)
        return false
      watchForOperation(key, operation)

      if (!watchCreated) {
        watchCreated = true
        estimatedTotalOperations.incrementAndGet()
      }
    }

    isCompletedByMe = operation.safeTryComplete()
    if (isCompletedByMe)
      return true

```

```safeTryComplete()``` is executed after updating ```watchKey```. Hence, it is possible that the lock of this request is held by **another thread**. The deadlock happens if this ```tryCompleteElseWatch``` is holding the **lock** required by **another thread**.

It seems to me the simple approach is to remove ```operation.safeTryComplete```. That should be fine since we have called ```tryComplete``` before.

Reply:
@chia7712 : 

1. I think we still need `operation.safeTryComplete` in `DelayedOperation.tryCompleteElseWatch()`. The reason is that after the `operation.tryComplete()` call, but before we add the key to watch, the operation could have been completed by another thread. Since that thread doesn't see the registered key, it won't complete the request. If we don't call `operation.safeTryComplete` after adding the key for watch, we could have missed the only chance for completing this operation.

2. I am not sure if there is a deadlock caused by TransactionStateManager. I don't see updateCacheCallback hold any lock on stateLock. The following locking sequence is possible through TransactionStateManager.

  thread 1 : hold readLock of stateLock, call ReplicaManager.appendRecords, call tryCompleteElseWatch, hold lock on delayedOperation

  thread 2: hold lock on delayedOperation, call delayedOperation.onComplete, call removeFromCacheCallback(), hold readLock of stateLock.

However, since both threads hold readLock of stateLock, there shouldn't be a conflict.

Do you see the test fail due to a deadlock?

Reply:
> Do you see the test fail due to a deadlock?

the following read/write lock is from```stateLock``` of ```TransactionStateManager```

1. Thread_1: holding readlock and waiting for lock of delayed op (TransactionStateManager#appendTransactionToLog)
2. Thread_2: waiting for writelock (```TransactionCoordinatorConcurrencyTest#testConcurrentGoodPathWithConcurrentPartitionLoading```)
    ```
    val t = new Thread() {
      override def run(): Unit = {
        while (keepRunning.get()) {
          txnStateManager.addLoadingPartition(numPartitions + 1, coordinatorEpoch)
        }
      }
    }
    private[transaction] def addLoadingPartition(partitionId: Int, coordinatorEpoch: Int): Unit = {
      val partitionAndLeaderEpoch = TransactionPartitionAndLeaderEpoch(partitionId, coordinatorEpoch)
      inWriteLock(stateLock) {
        loadingPartitions.add(partitionAndLeaderEpoch)
      }
    }
    ```
3. Thread_3: holding lock of delayed op and waiting for readlock (another thread is trying to complete delayed op)

**deadlock**
1. Thread_1 is waiting for thread_3
1. Thread_3 is waiting for Thread_2
1. Thread_2 is waiting for thread_1

Reply:
@chia7712 : Thanks for the explanation. `stateLock` is created as an unfair ReentrantReadWriteLock. So, in that case, will thread_3's attempt for getting the readLock blocked after thread_2? Did the test actually failed because of this?

Reply:
> will thread_3's attempt for getting the readLock blocked after thread_2?

To the best of my knowledge, writers have preference over readers in order to avoid starvation. That behavior is not public and we can get some evidence from source code. for example:

```java
        final boolean readerShouldBlock() {
            /* As a heuristic to avoid indefinite writer starvation,
             * block if the thread that momentarily appears to be head
             * of queue, if one exists, is a waiting writer.  This is
             * only a probabilistic effect since a new reader will not
             * block if there is a waiting writer behind other enabled
             * readers that have not yet drained from the queue.
             */
            return apparentlyFirstQueuedIsExclusive();
        }
```

At any rate, the non-fair mode does not guarantee above situation does not happen. Hence, it would be better to avoid potential deadlock caused by ```tryCompleteElseWatch```.

> I think we still need operation.safeTryComplete in DelayedOperation.tryCompleteElseWatch()

How about using ```tryLock``` in tryCompleteElseWatch? It avoids conflicting locking and still check completion of delayed operations after adding watches?

Reply:
@chia7712 : Thanks for the explanation. I agree that it's a potential problem.

Does using `tryLock` in `tryCompleteElseWatch()` lead us back to the previous issue that we could miss the opportunity to to complete an operation (fixed with KAFKA-6653)?

Another possibly is that we hold the lock in delayed operation while adding the operation to watch list and do the final `safeTryComplete()` check. This way, when the delayed operation is exposed to another thread, every thread, including the caller, always first acquires the lock in delayed operation. This should avoid all potential deadlocks between `tryCompleteElseWatch()` and `checkAndComplete()`. What do you think?

Reply:
> Another possibly is that we hold the lock in delayed operation while adding the operation to watch list and do the final safeTryComplete() check. This way, when the delayed operation is exposed to another thread, every thread, including the caller, always first acquires the lock in delayed operation. This should avoid all potential deadlocks between tryCompleteElseWatch() and checkAndComplete()

nice idea. I have addressed this approach.

Reply:
Do we still need this change to avoid deadlocks?

Reply:
Yep. if we add it to watch list too early, the concurrent issue happens as tryCompleteElseWatch calls tryComplete without locking.

Reply:
@junrao This approach can not resolve all potential deadlock. For example:

1. thread_a gets lock of op
1. thread_a adds op to watch list
1. thread_a calls op#tryComplete (and it requires lock_b)
1. thread_b holds lock_b
1. thread_b sees op from watch list
1. thread_b needs lock of op

Hence, we are facing the following issues.

1. the last completion check can cause deadlock after the op is exposed to other threads (by watch list).
1. the last completion check can not be removed due to KAFKA-6653.

How about using ActionQueue to resolve it? We add completion check to ActionQueue after adding op to watch list. All handlers are able to complete it even if they don't have same key.

Reply:
@chia7712 : What you described is possible, but probably not an issue in practice. Since tryComplete() always completes a delayed operation asynchronously, there is no reason for the caller of a delayed operation to hold any lock while calling tryComplete. Therefore, in step 4 above, the first lock that thread_b (assuming it's well designed) can acquire should be the lock in delayed operation.

Reply:
> there is no reason for the caller of a delayed operation to hold any lock while calling tryComplete. 

You are right. However, not sure how to maintain that "well designed" code in the future as the deadlock is implicit. It seems to me avoiding multiples exclusive lockings can avoid the deadlock.

I will keep current approach since the story I described may be overkill in practice.

Reply:
Is this change still necessary now that we always call tryComplete() with lock in tryCompleteElseWatch?